### PR TITLE
support kubeconfig env

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

### What is changed and how does it work?
Add support for KUBECONFIG env when starting components.
This env enables running and debugging this operator in local development environment easier.

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Helm charts change
 - Has Go code change yes
 - Has CI related scripts change
 - Has documents change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note

 ```

### CODE REVIEW

**Comment by yeya24** [comment]
Looks like the ci met the problem of `failed to watch file "/var/log/pods/15de0014-b112-11e9-b349-02427567777a/tidb-operator-e2e/0.log": no space left on device++`

**Comment by weekface** [comment]
Yes, I rebuild the CI env, and will try again later.

File: cmd/discovery/main.go [hunk 1]
```diff
@@ -38,10 +39,23 @@ func main() {
...
+		cfg, err = clientcmd.BuildConfigFromFlags("", kubeConfigFile)
+		if err != nil {
+			glog.Fatalf("failed to create config from file: %s %v\n", kubeConfigFile, err)
+		}
+	} else {
+		cfg, err = rest.InClusterConfig()
+		if err != nil {
+			glog.Fatalf("failed to get config: %v", err)
+		}
 	}
```
Comment (onlymellb) [inline cmd/discovery/main.go]: Can you extract this logic into a method `NewKubeCli(kubeCfg string) kubernetes.Interface` and then reference this method to initialize the k8s client? In addition, the client of the following CR `versioned.NewForConfig` should be initialized in the same way.

Comment (yeya24) [reply cmd/discovery/main.go]: I have get the logic into a new method which return *rest.config

**Comment by yeya24** [comment]
Sorry for the delay. Could take a look? @weekface @onlymellb Thanks!

**Comment by weekface** [comment]
@aylei This PR just add the kubeconf env support to operator. I think we can merge it and extract these logics in another PR.

**Comment by yeya24** [comment]
@weekface @aylei  Should I extract the logics from this PR and these initialization in tests? If I need to, I think I can work on it this weekend.

**Comment by tennix** [comment]
I don't think we can run the controller outside of the cluster, the controller needs to access PD by service name, the service name is not resolvable if not in a pod.

**Comment by yeya24** [comment]
> I don't think we can run the controller outside of the cluster, the controller needs to access PD by service name, the service name is not resolvable if not in a pod. But other components should work IIRC.
Make sense. Maybe it is better to close this PR.

**Pull request closed without merging**